### PR TITLE
[columnar] fix create table as parallelism bug

### LINKED ISSUE #50: Parallelism can block creation of columnar tables or views
Example:

```sql
CREATE TABLE sampletable (x numeric);
INSERT INTO sampletable SELECT i FROM generate_series(1,1000000) i;

-- since there are some parallel workers reading from heap, and CommandCounterIncrement
-- is not expected to be called during parallel operations
CREATE TABLE sampletable_columnar USING columnar AS SELECT * FROM sampletable ORDER BY 1 ASC;
ERROR: cannot start commands during a parallel operation
```

In this case the workaround is to temporarily disable parallelism, e.g. `set max_parallel_workers = 1`.

### PR DESCRIPTION
**Pull request by JerrySievert**
This fixes #50.

This is not the solution that I wanted to write, but it is _a_ solution and it works.

Access to a hook that gives us the plan for the `CREATE TABLE` portion of a `CREATE TABLE foo AS` statement is available in the utility hook, but that hook is fired _after_ the table is created and the insert is finished.  A planner hook has access to the `SELECT` statement, but does not show the `CREATE TABLE` portion of the plan.

I searched for an alternate method of getting access at the correct time to mutate the plan, but after posting on the postgres slack, discord, and pgsql-hackers mailing list, there was not an answer.  If that somehow changes in the future, I will happily revert this for a much better solution.

For now, this is what it does:

* creates a planner hook
* if the query is of type `CMD_SELECT` it creates a lowercase copy of the query string
* it then searches the lower case query string for `create`, `table`, and `as` in that order
* if all are found, then it sets the parallelism to 0 (no parallel) and returns the plan

it is possible to trigger this with a query like: `SELECT "create", "table", "as", FROM foo` or some other permutation of this, but the likelihood of collision is fairly low.

as noted, it's not a perfect fix, but it at least fixes the issue.

### CODE REVIEW

**Review by wuputah — COMMENTED** [review]
Yeah, parsing the query string is certainly not ideal.

I believe this bug can occur on a `insert into ... select` statement as well - though not documented in #50, that may be a more common issue than `create table ... select`.

**Comment by JerrySievert** [comment]
unfortunately, I've found no other path to fixing this.  if it also fails on an `insert` then we can blacklist that as well.

as I noted, I'm happy to fix this in a correct way if ever there is a correct way, but as of now, there doesn't appear to be a way to connect them, as even getting the `queryId` of the create statement occurs after the table is created.  there are only a couple of hooks that allow for plan mutation or even access to the `Query *` itself, which is unfortunate.

until someone comes forward with a solution that I've not found (I've combed through postgres code repeatedly), I'm not convinced there's another solution.

that said, if you can create a test case that still fails with this patch, I'm happy to include it, at least until we can toss this code out with a better solution.

**Comment by JerrySievert** [comment]
adding that it does not fail on insert, nothing to do with these code changes.

File: columnar/src/backend/columnar/columnar_customscan.c [hunk 1]
```diff
@@ -302,6 +311,79 @@ columnar_customscan_init()
...
+
+	/*
+	 * In the case of a CREATE TABLE AS query, we are not able to successfully
+	 * drop out of a parallel insert situation.  This checks for a CMD_SELECT
+	 * and in that case examines the query string to see if it matches the
+	 * pattern of a CREATE TABLE AS.  If so, set the parallelism to 0 (off).
+	 *
+	 * We are using strcasestr, which should give us
+	 */
+	if (parse->commandType == CMD_SELECT)
```
Comment (mkaruza) [inline columnar/src/backend/columnar/columnar_customscan.c:376]: It is fine to write into heap when parallel select will be planned. We should check if relation that is going to be written is columnar or something else.

**Review by mkaruza — CHANGES_REQUESTED** [review]

**Comment by mkaruza** [comment]
This issue related to `CommandCounterIncrement` function that is used when we write heap metadata tables. Maybe solution to fix this can be addressed there. If not, although not prettiest solution, should be good enough to mitigate this problem.

File: columnar/src/backend/columnar/columnar_customscan.c [hunk 2]
```diff
@@ -302,6 +311,79 @@ columnar_customscan_init()
...
+	PlannedStmt *stmt;
+
+	if (PreviousPlannerHook)
+	{
+		stmt = (*PreviousPlannerHook)(parse, query_string, cursorOptions, boundParams);
+	}
+	else
+	{
+		stmt = standard_planner(parse, query_string, cursorOptions, boundParams);
+  }
```
[the remark this answers is not in the document]
Comment (JerrySievert) [reply columnar/src/backend/columnar/columnar_customscan.c:366]: missed that one, thought I got them all.

File: columnar/src/backend/columnar/columnar_customscan.c [hunk 3]
```diff
@@ -302,6 +311,79 @@ columnar_customscan_init()
 	RegisterCustomScanMethods(&ColumnarScanScanMethods);
 }
 
+static bool IsCreateTableAs(const char *query)
```
[the remark this answers is not in the document]
Comment (JerrySievert) [reply columnar/src/backend/columnar/columnar_customscan.c:314]: missed that one too, good catch,

**Comment by JerrySievert** [comment]
agreed that the problem should be addressed around `CommandCounterIncrement` and was one of my earliest attempts.  unfortunately, it still gets called later, and only gets called during the parallel select.

File: columnar/src/backend/columnar/columnar_customscan.c (hunk 1 above)
Comment (JerrySievert) [reply columnar/src/backend/columnar/columnar_customscan.c:376]: we only have access to what (if anything) is being written via the utility hook, that is fired off _after_ the data is written.  thus we cannot determine the type of table being written.  but, the side effect is that parallelism is turned off for a select into a write, which occurs in heap anyways.

**Review by wuputah — APPROVED** [review]
until such time as a better solution present itself, I'm OK with shipping this; we can always change it later.

File: columnar/src/backend/columnar/columnar_customscan.c (hunk 1 above)
Comment (mkaruza) [reply columnar/src/backend/columnar/columnar_customscan.c:378]: If we are disabling parallel CTAS and can't filter `into` table please put comment somewhere in code about this change.

File: columnar/src/backend/columnar/columnar_customscan.c [hunk 4]
```diff
@@ -302,6 +311,81 @@ columnar_customscan_init()
...
+
+	/*
+	 * In the case of a CREATE TABLE AS query, we are not able to successfully
+	 * drop out of a parallel insert situation.  This checks for a CMD_SELECT
+	 * and in that case examines the query string to see if it matches the
+	 * pattern of a CREATE TABLE AS.  If so, set the parallelism to 0 (off).
+	 *
+	 * We are using strcasestr, which should give us
+	 */
+	if (parse->commandType == CMD_SELECT)
```
Comment (mkaruza) [inline columnar/src/backend/columnar/columnar_customscan.c:378]: If we are disabling parallel CTAS and can't filter into table please put comment somewhere in code about this change. (also the upper comment is missing its end)

**Review by mkaruza — CHANGES_REQUESTED** [review]

File: columnar/src/backend/columnar/columnar_customscan.c (hunk 4 above)
Comment (JerrySievert) [reply columnar/src/backend/columnar/columnar_customscan.c:383]: pushed with updated comments.

we can filter into the table.  not sure what other comment you'd want about parallel ctas other than the fact that we're explicitly disabling parallelism for it?

**Review by mkaruza — APPROVED** [review]

**Pull request merged**